File: mobile/src/main/java/info/papdt/express/helper/ui/ScannerActivity.java
package info.papdt.express.helper.ui;

import android.Manifest;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.support.design.widget.Snackbar;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.view.View;
import android.view.WindowManager;

import com.google.zxing.Result;

import info.papdt.express.helper.R;
import info.papdt.express.helper.ui.common.AbsActivity;
import me.dm7.barcodescanner.zxing.ZXingScannerView;

public class ScannerActivity extends AbsActivity implements ZXingScannerView.ResultHandler {

	private ZXingScannerView mScannerView;

	public static final int REQUEST_CODE_SCAN = 10001, RESULT_GET_RESULT = 1000;
	public static final String EXTRA_RESULT = "extra_result";

	private static final int REQUEST_PERMISSION = 20001;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
			getWindow().getDecorView().setSystemUiVisibility(
					View.SYSTEM_UI_FLAG_LAYOUT_STABLE | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
			);
			getWindow().addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION);
			getWindow().setStatusBarColor(Color.TRANSPARENT);
		}
		setContentView(R.layout.activity_scanner);

		if (!isCameraPermissionGranted()) {
			if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.CAMERA)) {
				$(R.id.iv_error).setVisibility(View.VISIBLE);
				new AlertDialog.Builder(this)
						.setTitle(R.string.dialog_explanation_permission_title)
						.setMessage(R.string.dialog_explanation_permission_message)
						.setPositiveButton(R.string.dialog_explanation_permission_pos_btn, new DialogInterface.OnClickListener() {
							@Override
							public void onClick(DialogInterface dialogInterface, int i) {
								jumpToSettings();
							}
						})
						.setNegativeButton(android.R.string.cancel, null)
						.show();
			} else {
				ActivityCompat.requestPermissions(
						this,
						new String[]{Manifest.permission.CAMERA},
						REQUEST_PERMISSION
				);
			}
		}
	}

	@Override
	protected void setUpViews() {
		mScannerView = $(R.id.scanner_view);
		mActionBar.setDisplayHomeAsUpEnabled(true);
	}

	@Override
	protected void onResume() {
		super.onResume();
		if (isCameraPermissionGranted()) {
			startScan();
		}
	}

	@Override
	public void onPause() {
		super.onPause();
		mScannerView.stopCamera();
	}

	@Override
	public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
		if (requestCode == REQUEST_PERMISSION) {
			if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
				startScan();
			} else {
				Snackbar.make(mScannerView, R.string.toast_permission_denied, Snackbar.LENGTH_INDEFINITE)
						.setAction(R.string.toast_permission_denied_action, new View.OnClickListener() {
							@Override
							public void onClick(View view) {
								jumpToSettings();
							}
						})
				.show();
			}
		}
	}

	private void startScan() {
		$(R.id.iv_error).setVisibility(View.INVISIBLE);
		mScannerView.setResultHandler(this);
		mScannerView.startCamera();
	}

	private boolean isCameraPermissionGranted() {
		return ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED;
	}

	private void jumpToSettings() {
		Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS, Uri.parse("package:" + getApplication().getPackageName()));
		startActivity(intent);
	}

	@Override
	public void handleResult(Result result) {
		Intent intent = new Intent();
		intent.putExtra(EXTRA_RESULT, result.getText());
		setResult(RESULT_GET_RESULT, intent);
		finish();
	}

}

File: mobile/src/main/kotlin/info/papdt/express/helper/ui/ScannerActivity.kt
package info.papdt.express.helper.ui

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.provider.Settings
import android.view.Menu
import android.view.MenuItem
import com.google.android.material.snackbar.Snackbar
import androidx.core.content.ContextCompat
import android.view.View

import info.papdt.express.helper.ui.common.AbsActivity
import me.dm7.barcodescanner.zxing.ZXingScannerView
import moe.feng.kotlinyan.common.*
import android.graphics.BitmapFactory
import android.graphics.Bitmap
import com.google.zxing.*
import com.google.zxing.common.HybridBinarizer
import info.papdt.express.helper.R
import info.papdt.express.helper.support.SnackbarUtils
import java.io.IOException
import java.lang.Exception
import java.util.*

class ScannerActivity : AbsActivity(), ZXingScannerView.ResultHandler, PermissionActivity {

	companion object {

		const val REQUEST_CODE_SCAN = 30001
		const val EXTRA_RESULT = "extra_result"

		private const val REQUEST_PERMISSION = 20001
		private const val REQUEST_CODE_GALLERY = 20002

		const val ACTION_SCAN_TO_ADD = "info.papdt.express.helper.ACTION_SCAN_TO_ADD"

	}

	private val scannerView: ZXingScannerView
            by lazyFindNonNullView(R.id.scanner_view)

	private lateinit var multiFormatReader: MultiFormatReader

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
        window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or
                View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
        window.statusBarColor = Color.TRANSPARENT
		setContentView(R.layout.activity_scanner)

		val hints = EnumMap<DecodeHintType, Any>(DecodeHintType::class.java)
		hints[DecodeHintType.POSSIBLE_FORMATS] = scannerView.formats
		multiFormatReader = MultiFormatReader()
		multiFormatReader.setHints(hints)

		runWithPermission(Manifest.permission.CAMERA) { startScan() } ?: run {
			buildAlertDialog {
				titleRes = R.string.dialog_explanation_permission_title
				messageRes = R.string.dialog_explanation_permission_message
				positiveButton(R.string.dialog_explanation_permission_pos_btn)  { _, _ -> jumpToSettings() }
				cancelButton { _, _ ->
					finish()
				}
				setOnCancelListener {
					finish()
				}
			}.show()
		}
	}

	override fun setUpViews() {

	}

	override fun onResume() {
		super.onResume()
		if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
				== PackageManager.PERMISSION_GRANTED) {
			startScan()
		}
	}

	override fun onPause() {
		super.onPause()
		scannerView.stopCamera()
	}

	override fun onCreateOptionsMenu(menu: Menu): Boolean {
		menuInflater.inflate(R.menu.menu_scanner, menu)
		return super.onCreateOptionsMenu(menu)
	}

	override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
		R.id.action_choose_from_gallery -> {
			val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
			intent.addCategory(Intent.CATEGORY_OPENABLE)
			intent.type = "image/*"
			try {
				startActivityForResult(intent, REQUEST_CODE_GALLERY)
			} catch (e: Exception) {
				buildAlertDialog {
					titleRes = R.string.dialog_no_documents_app_title
					messageRes = R.string.dialog_no_documents_app_summary
					okButton()
					neutralButton(R.string.dialog_no_documents_app_why_i_will_meet_this) { _, _ ->
						startActivity(Intent(Intent.ACTION_VIEW)
								.setData(Uri.parse(getString(R.string.broken_api_post_url))))
					}
				}.show()
			}
			true
		}
		else -> super.onOptionsItemSelected(item)
	}

	override fun onRequestPermissionsResult(requestCode: Int,
											permissions: Array<String>,
											grantResults: IntArray) {
		handleOnRequestPermissionsResult(requestCode, permissions, grantResults) {
			if (Manifest.permission.CAMERA == it) {
				Snackbar.make(
						scannerView,
						R.string.toast_permission_denied,
						Snackbar.LENGTH_SHORT
				).setAction(R.string.toast_permission_denied_action) { jumpToSettings() }.show()
			}
		}
	}

	override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) = ui {
		when (requestCode) {
			REQUEST_CODE_GALLERY -> {
				if (resultCode == RESULT_OK) {
					val uri = data?.data
					if (uri != null) {
						val result = asyncIO {
							val bitmap = getBitmapFromUri(uri)
							val pixels = IntArray(bitmap.width * bitmap.height) { 0 }
							bitmap.getPixels(
									pixels, 0,
									bitmap.width, 0, 0,
									bitmap.width, bitmap.height)
							bitmap.recycle()
							val src = RGBLuminanceSource(bitmap.width, bitmap.height, pixels)
							val binBmp = BinaryBitmap(HybridBinarizer(src))
							try {
								multiFormatReader.decodeWithState(binBmp)
							} catch (e: Exception) {
								null
							}
						}.await()
						if (result != null) {
							handleResult(result)
						} else {
							SnackbarUtils.makeInCoordinator(this,
									R.string.toast_cannot_find_any_barcode).show()
						}
					}
				}
			}
		}
	}

	@Throws(IOException::class)
	private fun getBitmapFromUri(uri: Uri): Bitmap {
		val pfd = contentResolver.openFileDescriptor(uri, "r") ?: throw IOException()
		val fileDescriptor = pfd.fileDescriptor
		val image = BitmapFactory.decodeFileDescriptor(fileDescriptor)
		pfd.close()
		return image
	}

	private fun startScan() {
		scannerView.setResultHandler(this)
		scannerView.startCamera()
	}

	private fun jumpToSettings() {
		val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
				Uri.parse("package:" + application.packageName))
		startActivity(intent)
	}

	override fun handleResult(result: Result) {
		val intent = Intent()
		intent[EXTRA_RESULT] = result.text
		setResult(RESULT_OK, intent)
		finish()
	}

}
